### ISSUE TITLE
jsondump doesn't correctly encode zero-sized arrays.

### ISSUE THREAD

**Issue by SiegeLordEx** [stan-dev/cmdstanpy#203]
#### Summary:
When passing zero size arrays (any of the dimensions are 0) as data, cmdstanpy loses the shape information.

#### Description:
Consider this code:

```python
import os
import cmdstanpy
import numpy as np

code = """
data {
  int num_points;
  matrix[num_points, 1] features;
  int labels[num_points];
}
parameters {
  vector[1] x;
}
model {
  x[1] ~ normal(0, 1);
}
generated quantities {
  real test_log_likelihood;
  test_log_likelihood = bernoulli_logit_lpmf(labels | features * x);
}
"""

filename = '/tmp/stan_test.stan'
with open(filename, 'w') as f:
  f.write(code)

model = cmdstanpy.CmdStanModel(stan_file=filename)
model.compile()

out_dir = '/tmp/zero_stan_out'
os.makedirs(out_dir)
num_points = 0
print(model.sample(
    data={
        'num_points': num_points,
        'features': np.ones([num_points, 1]),
        'labels': np.zeros([num_points], dtype=np.int32)
    }, output_dir=out_dir).summary())
```

When I run it, I get a runtime error. Looking at the logs I see this:

```
Exception: mismatch in number dimensions declared and found in context; processing stage=data initializat
ion; variable name=features; dims declared=(0,1); dims found=(0)  (in '/tmp/stan_test.stan' at line 4)
```

If I hack cmdstanpy/util.py to always use `rdump` instead of `jsondump`, it works fine, so it seems that `jsondump` is a lossy encoding in this situation.

#### Additional Information

The use case is that I have a generic regression model where sometimes I have held out data I want to compute some quantity on. When I don't, I tried to represent it as an empty array, but that didn't work out.

As a workaround, I'll use some dummy test data for now.

#### Current Version:
0.0.8

**Comment by ahartikainen** [comment]
related: https://github.com/stan-dev/cmdstanpy/issues/121

**Comment by bob-carpenter** [comment]
@SiegeLordEx : Thanks for the careful bug report.  This is definitely something we need to fix for exactly the reasons outlined in the issue report.

**Comment by mitzimorris** [comment]
hi, tested on `cmdstanpy-0.8.0`, with latest CmdStan release - `cmdstan-2.22.1` -

model:
```
data {
  int num_points;
  matrix[num_points, 1] features;
  int labels[num_points];
}
parameters {
  vector[1] x;
}
model {
  x[1] ~ normal(0, 1);
}
generated quantities {
  real test_log_likelihood;
  test_log_likelihood = bernoulli_logit_lpmf(labels | features * x);
}
```

data:
```
{'num_points': 0, 'features': [], 'labels': []}
```

console output:

```
>>> model.sample(data=data,output_dir=out_dir)
INFO:cmdstanpy:start chain 1
INFO:cmdstanpy:start chain 2
INFO:cmdstanpy:finish chain 2
INFO:cmdstanpy:start chain 3
INFO:cmdstanpy:finish chain 1
INFO:cmdstanpy:start chain 4
INFO:cmdstanpy:finish chain 3
INFO:cmdstanpy:finish chain 4
CmdStanMCMC: model=multi_dim_data chains=4['method=sample', 'algorithm=hmc', 'adapt', 'engaged=1']
 csv_files:
	/Users/mitzi/github/stan-dev/cmdstanpy/test/data/zero_stan_out/multi_dim_data-202002031056-1.csv
	/Users/mitzi/github/stan-dev/cmdstanpy/test/data/zero_stan_out/multi_dim_data-202002031056-2.csv
	/Users/mitzi/github/stan-dev/cmdstanpy/test/data/zero_stan_out/multi_dim_data-202002031056-3.csv
	/Users/mitzi/github/stan-dev/cmdstanpy/test/data/zero_stan_out/multi_dim_data-202002031056-4.csv
 console_files
	/Users/mitzi/github/stan-dev/cmdstanpy/test/data/zero_stan_out/multi_dim_data-202002031056-1.txt
	/Users/mitzi/github/stan-dev/cmdstanpy/test/data/zero_stan_out/multi_dim_data-202002031056-2.txt
	/Users/mitzi/github/stan-dev/cmdstanpy/test/data/zero_stan_out/multi_dim_data-202002031056-3.txt
	/Users/mitzi/github/stan-dev/cmdstanpy/test/data/zero_stan_out/multi_dim_data-202002031056-4.txt
```

sampler runs - throws exception in `generated_quantities` block
```
Exception: multiply: m1 must have a positive size, but is 0; dimension size expression = rows() (in '/Users/mitzi/github/stan-dev/cmdstanpy/test/data/multi_dim_data.stan', line 14, column 2 to column 68)
```

the samples csv files all have `nan` value for variable `test_log_likelihood`

```
lp__,accept_stat__,stepsize__,treedepth__,n_leapfrog__,divergent__,energy__,x.1,test_log_likelihood
# Adaptation terminated
# Step size = 0.940802
# Diagonal elements of inverse mass matrix:
# 1.06797
-0.028936,0.973392,0.940802,1,3,0,0.629099,0.240566,nan
-0.343467,0.940549,0.940802,1,3,0,0.363895,0.828815,nan
-0.525908,0.903503,0.940802,2,3,0,1.30974,1.02558,nan
```

question to @SiegeLordEx  - is this error message helpful?  this is running against the new stanc3 compiler

**Comment by SiegeLordEx** [comment]
I'm not sure whether `m1` refers to first or second dimensions of `features`. I don't expect [0, N] * [N] to raise an error. Would it still raise an error if features were `np.zeros([0, 1])`? The `NaN`s are fine, although I think 0 can also be the "correct" answer, as we're doing a sum(lpm), and sum has 0 as an identity which is a reasonable value for zero-size arrays.

**Comment by bob-carpenter** [comment]
Sum should take value 0 for empty inputs and product value 1, but I don't see the relevance here.

Multiplying an object of shape matrix[0, N] by an object of shape vector[N] should produce an object of shape vector[0], so there's nothing to sum.

I'm not sure it'll work in the languauge as is, but if it doesn't work, it's a bug.

**Comment by SiegeLordEx** [comment]
> Sum should take value 0 for empty inputs and product value 1, but I don't see the relevance here.

Sorry, I wasn't clear. As far as I can tell, the `NaN` comes about in @mitzimorris's code from this line `bernoulli_logit_lpmf(labels | features * x)`, which I assumed for vector-valued inputs does a sum reduction, hence me mentioning it.

**Comment by bob-carpenter** [comment]
> On Feb 3, 2020, at 1:20 PM, Pavel Sountsov <notifications@github.com> wrote:
> 
> Sum should take value 0 for empty inputs and product value 1, but I don't see the relevance here.
> 
> Sorry, I wasn't clear. As far as I can tell, the NaN comes about in @mitzimorris's code from this line bernoulli_logit_lpmf(labels | features * x), which I assumed for vector-valued inputs does a sum reduction, hence me mentioning it.

Yes, it does, but if everything's size zero the desnity is 1 (it's a size-zero product) and thus ehte log density should be 0 (being a size-zero sum and also matching the log of the density.  

The program

  int z[0];
  real x[0];
  print("bllpmf = ", bernoulli_logit_lpmf(z | x ));

prints 0 as it should.

**Comment by mitzimorris** [comment]
> I'm not sure whether `m1` refers to first or second dimensions of `features`. 

the error message:
```
... multiply: m1 must have a positive size, but is 0; dimension size expression = rows()...
```
comes from the stan math library function for matrix multiplication A * B (of doubles) -
`m1` refers to matrix A, i.e., the variable `features` and the error message is saying that the rows dimension of `features` is size 0 but must be positive.


>  Multiplying an object of shape matrix[0, N] by an object of shape vector[N] should produce an object of shape vector[0]  .... I'm not sure it'll work in the languauge as is, but if it doesn't work, it's a bug.
> […](#)

OK, it's a bug in the math lib.

closing this issue.

**Comment by mitzimorris** [comment]
filed https://github.com/stan-dev/math/issues/1669
